src/main/java/io/supertokens/utils/Utils.java

```diff
@@ -84,6 +84,16 @@ public static String hashSHA256(String base) throws NoSuchAlgorithmException {
         return MessageDigest.getInstance("SHA-256").digest(base);
     }
 
+    public static String hashSHA256Base64UrlSafe(byte[] base) throws NoSuchAlgorithmException {
+        byte[] hashBytes = Utils.hashSHA256Bytes(base);
+        return Base64.getUrlEncoder().encodeToString(hashBytes);
+    }
+
+    public static String hashSHA256Base64(byte[] base) throws NoSuchAlgorithmException {
+        byte[] hashBytes = Utils.hashSHA256Bytes(base);
+        return Base64.getEncoder().encodeToString(hashBytes);
+    }
+
```

Comment:
In my mind, extracting those into functions entail creating helper functions for:
- deviceId -> deviceIdHash
- (deviceId + userInputCode) -> linkCode
- linkCode -> linkCodeHash

And then these functions would be used everywhere as and when needed. This way, we have one central place where do do conversations using base64 and bytes etc..

Reply:
My preference would be to do the b64 conversions at the API level. It doesn't matter much, but passing the unparsed version around and doing the conversion multiple times doesn't feel clean.